zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,408 @@
+use halo2_proofs::plonk::{ConstraintSystem,Circuit, Error};
+use halo2_proofs::circuit::{Layouter,SimpleFloorPlanner,Value};
+use halo2_proofs::halo2curves::ff::PrimeField;
+use halo2wrong_integer::{IntegerConfig, rns::Rns, IntegerChip, IntegerInstructions, rns::Integer, UnassignedInteger, Range, AssignedInteger};
+use halo2wrong_maingate::{RangeChip, MainGate, RegionCtx};
+
+/// THIS CIRCUIT PERFORMS A NON-NATIVE FUNCTION EVAL USING BARYCENTRIC-FORMULA
+/// P(X) = (const) * sum (d_i * (w^i/x-w^i)) where const = (x^N - 1)/N with w^N = 1 and X^N != 1
+
+#[derive(Clone, Debug)]
+pub struct BlobKZGConfig {
+    integer_config: IntegerConfig
+}
+
+#[derive(Debug)]
+pub struct BlobKZGCircuit<W: PrimeField, N: PrimeField>{ // W is the wrong field, N is the native field
+    datapoints: [W; 4096],
+    x: W, 
+    w: W,
+    n: W,
+    witness_values: WitnessValues<W>,
+    _marker: std::marker::PhantomData<N>
+}
+
+impl<W: PrimeField, N:PrimeField> Default for BlobKZGCircuit<W,N> {
+    fn default() -> Self {
+        BlobKZGCircuit { datapoints: [W::default();4096], x: W::default(), w: W::default(), n: W::default(), witness_values: WitnessValues::<W>::default(), _marker: std::marker::PhantomData }
+    }
+}
+#[derive(Clone, Debug)]
+struct WitnessValues<F: PrimeField>{
+    evals: [F;4098],
+    wis: [F;4096],
+    xns: [F;257],
+    x2is: [F;256],
+    nbits: [F;256] 
+}
+impl<F: PrimeField> Default for WitnessValues<F> {
+    fn default() -> Self {
+        WitnessValues { evals: [F::default();4098], wis: [F::default();4096], xns: [F::default();257], x2is: [F::default();256], nbits: [F::default();256] }
+    }
+}
+impl<W: PrimeField> WitnessValues<W> {
+    fn new(x:&W, n:&W, w:&W, datapoints: &[W; 4096]) -> WitnessValues<W> {
+        let x2is: [W; 256] = {
+            let mut x2is = [W::default(); 256];
+            let mut state = *x;
+            for x in &mut x2is {
+                *x = state;
+                state *= state;
+            }
+            x2is
+        };
+        let nbits: [W; 256] = {
```

Comment:
Unfortunately, I do not no Halo2Wrong API and do not sure that I have correctly understood the related parts. However, if they work as I suppose and bit factorization of n (log2 of the order of w) is predefined, than I think that it should work correctly. This part of the code (the lines 54-71) can be written in a shorter way (defining nbits[j*8+i] directly without the boolean array), but it is not a problem.